### DISCUSSION TITLE
Replace `isPaused` state with runtime check with `rps`

### DISCUSSION THREAD

**Discussion by smol-ninja** [sablier-labs/flow#271] [Ideas]
A paused stream is characterized by two states that are analogous to each other: `rps == 0` and `isPaused == true`. Thus we can remove  `isPaused` state from the contract and instead return `true` if `rps == 0`.

### Pros
1. 2500 gas saving in `paused`, `void` and `restart`
2. Single state to track whether the stream is paused or not. Currently we have to mandatory update both `rps` and `isPaused`

### Cons
1. None

> [!NOTE]
>
> As we're quite ready for audits with flow, we can discuss this for an upcoming version of the protocol, post-launch. However, this change is minor so, if we all agree, it can be included before the audit. I am in favour of including this before audit.

cc @sablier-labs/solidity

**Comment by andreivladbrg** [comment]
I am against this proposal for the same reason why we intially introduced `isStream` in lockup contracts, and we no longer check for `sender != address(0)` if a stream exists. The same logic applies to more recent `wasCanceled` and `isDepleted`.

Using a dedicated boolean is a best practice for checking entry points in functions and ensures clarity and security.

> 2500 gas saving in paused, void and restart

In this case, I prioritize a robust and secure pattern over gas optimizations. But if you want to optimize we can, instead of removing `isPaused`, we could remove updating `rps` to zero in `pause` /`void` (since we check `isPaused = true` in `ongoingDebt` function). Tho, I’m not sure if I agree with this approach either, because updating it to zero is more secure.

**Reply by smol-ninja** [reply]
> Using a dedicated boolean is a best practice for checking entry points in functions and ensures clarity and security

True for Lockup but now for Flow. The case with lockup is different than flow. Flow has a concept of `rps` which Lockup does not. 

> I I prioritize a robust and secure pattern over gas optimizations

How would it be less secure?

> we could remove updating rps to zero in pause /void

Using `rps` as 0 for paused stream is very intuitive implementation.

**Reply by smol-ninja** [reply]
> robust and secure pattern over gas optimizations

I think you contradicted yourself against [this arguement](https://github.com/sablier-labs/flow/discussions/269#discussioncomment-10764341) where you are prioritising gas efficiency over a robust pattern.

**Reply by PaulRBerg** [reply]
To @andreivladbrg 

> why we intially introduced `isStream` in lockup contracts

No, that was a different problem situation.

**The issue in Lockup was that the stream creator could have chosen a zero address for the `sender` intentionally**. So there were two possibilities:

1. Default value of zero address for `sender`, meaning the stream DOES NOT EXIST
2. Intentional use of zero address for `sender`, meaning the stream entity SHOULD EXIST BUT IT DOES NOT because of the `sender != address(0)` check

Whereas in this case, Shub is right that `rps == 0` **if and only if** `status == PAUSED`. Thus, because these are isometric states, there's no point in keeping the additional `isPaused` boolean when we can check for `rps == 0` directly.

> updating it to zero is more secure.

Exactly. That's what it means for the stream to be paused.

---

To @smol-ninja:

> I think you contradicted yourself against https://github.com/sablier-labs/flow/discussions/269#discussioncomment-10764341

Andrei didn't contradict himself. He had a different rationale and it was a different context. As he said: `I prioritize a robust and secure pattern over gas optimizations`.

Still, his rationale is incorrect for a different reason (explained above).

> Using `rps` as 0 for paused stream is very intuitive implementation.

Indeed. It would be strange for the status to be `PAUSED` but also have a non-zero RPS.

**Reply by andreivladbrg** [reply]
correct, but actually we both did it

one discussion (269) is saying to have a robust way of updating `st` (in regard of gas costs), and this one is saying that let's remove the `isPaused` (which is a robust pattern) in favor of gas costs

---

so, the best option IMO now is to: 
- keep `isPaused`
- update `st` everytime `sd` is

**Reply by PaulRBerg** [reply]
> one discussion (269) is saying to have a robust way of updating st (in regard of gas costs), and this one is saying that let's remove the isPaused (which is a robust pattern) in favor of gas costs

Different problem situations. Let's focus on the specific explanations provided in each discussion.

**Reply by andreivladbrg** [reply]
Shub introduced 269 discussions into this, so I responded corresponding with the subject of the comment

**Reply by andreivladbrg** [reply]
>The issue in Lockup was that the stream creator could have chosen a zero address for the sender intentionally. So there were two possibilities:

that's incorrect, the stream creator could have not chosen a zero address, as we've had check if `sender == address(0)` and reverted. (also rps is an input param in `create` but, as we do with the sender, we check wether is 0 or not)

> Whereas in this case, Shub is right that rps == 0 if and only if status == PAUSED

given the above comment, it is the same thing with lockup, we have the invariant: `if stream exists then sender != address(0)`

**Reply by PaulRBerg** [reply]
> Shub introduced 269 discussions into this

Indeed. You should have just pointed out the error, not escalated it.

**Reply by PaulRBerg** [reply]
> the stream creator could have not chosen a zero address, as we've had check if sender == address(0) and reverted

Sorry, my wording was unclear.

What I meant to say was that the sender may have wanted to set the `sender` to `address(0)`. He was indeed blocked in the create function, but the point is that there was a logical possibility for the sender to the zero address and the stream to EXIST.

> also rps is an input param in create but, as we do with the sender, we check wether is 0 or not)

In Lockup, we had that check only when the boolean flag DID NOT exist.

In Flow, we have that check even if the boolean flag DOES exist. So the boolean flag is redundant.

**Reply by smol-ninja** [reply]
> On Lockup

`sender` used to be zero address [before this PR](https://github.com/sablier-labs/v2-core/pull/1018). That is why `isStream` was introduced in previous versions of Lockup. But now, I also agree to remove `isStream` from the Lockup and use runtime check for `sender` address.

> let's remove the `isPaused` (which is a robust pattern) in favor of gas costs

Gas cost saving is a consequence of this change. But that was not the primary motivation (I should have been clear about it in the OP). The primary reason is, as Paul mentioned, "isometric states". We have two contract states that are analogous to each other.

**Reply by PaulRBerg** [reply]
> I also agree to remove isStream from the Lockup and use runtime check for sender address.

I don't. Different problem situations because of the logical possibilities of users wanting to set `sender` to `address(0)`.

**Reply by smol-ninja** [reply]
Understood, though I'm not entirely convinced how the sender could set the address to 0 if that's no longer allowed. We can dive deeper into this through a separate Lockup discussion later.

@andreivladbrg @PaulRBerg if could you share your thoughts, if anything new, on whether we should proceed with this proposal? To clarify, my reasoning is that when a stream is paused, rps is 0 and `isPaused` is true. Conversely, when it's unpaused, rps is greater than 0 and `isPaused` is false. This means we're using two storage states to convey the same information.

**Reply by andreivladbrg** [reply]
I still believe that having dedicated booleans is a safer design choice (it also improves code clarity) and provides a more robust pattern in case we scale to a more complex version with additional rps changes. I consider the version from my [comment here](https://github.com/sablier-labs/flow/discussions/271#discussioncomment-10776215) to be the best.

Regarding the "isometry" point, we have these possible scenarios (when the stream does not exist or they are not in sync):

- stream **does not** exist: `rps = 0 && isPaused = false` 
- stream **does** exist: `rps > 0 && isPaused = false`
- stream **does** exist: `rps = 0 && isPaused = true`

> my reasoning is that when a stream is paused, rps is 0 and isPaused is true. Conversely, when it's unpaused, rps is greater than 0 and isPaused is false. This means we're using two storage states to convey the same information.

You’re correct. I never disagreed with this; I was simply discussing about "best practices"

> sender used to be zero address https://github.com/sablier-labs/v2-core/pull/1018. That is why isStream was introduced in previous versions of Lockup

I don’t think you have the full context, sorry. Back in 2022, we used to also check the sender before introducing the `isStream` boolean.

---

> Indeed. You should have just pointed out the error, not escalated it.

I didn’t find any error in the logic, so I responded on the subject. I apologize if it seemed like I was escalating; that wasn’t my intent. I was simply acknowledging that Shub was right in his statement, and we both lacked consistency in our thinking

re `isStream + sender` and `rps` analogy, that's exactly the same. 

for lockup:
1. we checked wether `sender == address(0)` in `create`
2. we checked wether a stream existed based on `sender == address(0)`

for flow:
1. we check wether `rps == 0` in [create](https://github.com/sablier-labs/flow/blob/8eb58ac5d8237944bac910075aa3238543df2442/src/SablierFlow.sol#L564-L567)
2. we will check wether a stream is paused based on `rps == 0`

> What I meant to say was that the sender may have wanted to set the sender to address(0). He was indeed blocked in the create function, but the point is that there was a logical possibility for the sender to the zero address and the stream to EXIST

as you can see above, that's pretty much the same thing

> In Lockup, we had that check only when the boolean flag DID NOT exist.

We also have this in the current version, although we now use the `isStream` bool

---

As far as I can tell, the majority is in favor of making this change, and I don’t have anything further to add. Let’s go ahead with the change.

**Reply by smol-ninja** [reply]
Thanks for sharing your views on this, Andrei. Appreciate it. Understood everything you said and agree with you. Just a small note:

> provides a more robust pattern in case we scale to a more complex version with additional rps changes

As far as I understand, we follow an adaptive decision-making approach when it comes to writing contracts. So, if we move towards a complex version of Flow with additional rps changes (as necessitated by the future requirement), we can then decide/discuss to implement the `isPaused` storage as needed. I will also be in favor of that.

**Reply by PaulRBerg** [reply]
Thank you for the additional color @andreivladbrg. I responded to the main point you made in the Loom recording shared below.

> I didn’t find any error in the logic, so I responded on the subject. I apologize if it seemed like I was escalating; that wasn’t my intent

Got it. No worries.

**Comment by PaulRBerg** [comment]
Agree with Shub's proposal, let's implement it.

**Comment by andreivladbrg** [comment]
Since we’re making this change, should we also remove the [isStream](https://github.com/sablier-labs/flow/blob/8eb58ac5d8237944bac910075aa3238543df2442/src/types/DataTypes.sol#L70) boolean? Or should this be discussed in a separate discussion? (IMO this would be the logical option)

**Reply by smol-ninja** [reply]
@PaulRBerg has some disagreement on that (https://github.com/sablier-labs/flow/discussions/271#discussioncomment-10776495).  But I'd recommend to discuss it separately in the Lockup repo.

**Reply by andreivladbrg** [reply]
> ut I'd recommend to discuss it separately in the Lockup repo.

we have `isStream` here as well, so we can discuss it here in this repo (separate discussion)

**Reply by smol-ninja** [reply]
Oh right. I am in favour of it as well and think that it would also work to compute `isStream` function using runtime check.

**Reply by smol-ninja** [reply]
Re-opened it as the discussion about `isStream` is still pending.

**Reply by andreivladbrg** [reply]
let's have a different one

**Reply by PaulRBerg** [reply]
Yeah we should have this discussion about `isStream` in Lockup

**Comment by smol-ninja** [comment]
### Issue
- https://github.com/sablier-labs/flow/issues/276

**Comment by PaulRBerg** [comment]
I am still in favor of Shub' proposal. @andreivladbrg I responded to your long comment above through this [Loom recording](https://www.loom.com/share/6d87e709df8a46b39d280107db9cd52f?sid=14ef5b97-5fe2-4d14-9ba7-bc42e1bc7978).

---

To @smol-ninja:

> we follow an adaptive decision-making approach

Indeed. Also called YAGNI.

> This means we're using two storage states to convey the same information.

Correct

> I'm not entirely convinced how the sender could set the address to 0 if that's no longer allowed.

It's a matter of _logical possibility_. It is possible to have a Lockup stream with a zero address sender if the stream creator wishes so (as in, logically; I know that our if check blocks this in practice). But it is NOT possible to have a Flow stream with an `rps == 0` that is NOT paused.

**That we don't allow `rps == 0` in the create function is a subjective preference. It's not an objective necessity.**

This makes me think — why not lift this requirement? Users may want to set up the streams in advance, and fund them later.

**Reply by andreivladbrg** [reply]
> **That we don't allow `rps == 0` in the create function is a subjective preference. It's not an objective necessity.**
> This makes me think — why not lift this requirement? Users may want to set up the streams in advance, and fund them later

hmm, interesting idea, can you open a different disscusion for this with what exactly you have in mind? it seems like a different topic

also, re. "fund them later" - if you mean to `deposit` later,  that's possible at the moment

**Reply by PaulRBerg** [reply]
I will open a separate discussion, yes. I meant `adjustRatePerSecondAndDeposit`.

Could you please confirm whether you have accepted the refutation I have provided in the Loom video, or not?

**Reply by andreivladbrg** [reply]
> I meant adjustRatePerSecondAndDeposit.

hmm, in that case why make 2 separate actions instead of one? from my understanding you are suggesting:
1. create with rps = 0
2. wait time
3. adjustRpsAndDeposit with `rps > 0`

instead of:
1. wait time
2. `createAndDeposit` with `rps > 0`

we can further discuss this in the upcoming discussion

---



> Could you please confirm whether you have accepted the refutation I have provided in the Loom video, or not?

your point was already confirmed by me in my comment (this is why i haven't said anything). perhaps you missed it?

<img width="500" alt="Screenshot 2024-10-01 at 12 35 44 AM" src="https://github.com/user-attachments/assets/3bdb0f21-7758-41fa-9855-d29eca3b1f5d">

nervertherless, i will say it again. $`isPaused = true \implies rps = 0`$ && $`isPaused = false \implies rps > 0`$ is a correct statement

**Reply by PaulRBerg** [reply]
In the Loom video, I refuted this part (unrelated to your screenshot):

<img width="400" alt="SCR-20240930-tvio" src="https://github.com/user-attachments/assets/7387ab3d-4a77-4c5a-93f6-4f66027978f6">

Could you please confirm whether you have understood my criticism and if yes, accept it or criticize it?

**Reply by andreivladbrg** [reply]
> Could you please confirm whether you have understood my criticism and if yes, accept it or criticize it?

your criticism is "`stream does not exist: rps = 0 && isPaused = false` is correct but it doesn't matter to isPaused"  i accept it, you are correct, it does not matter

**Reply by PaulRBerg** [reply]
Thank you. Point settled!

I will open a separate discussion for the `adjustRatePerSecondAndDeposit` idea.
